src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posix.c

```diff
@@ -109,6 +111,20 @@ static void fd_node_destroy(grpc_exec_ctx *exec_ctx, fd_node *fdn) {
   gpr_free(fdn);
 }
 
+static void fd_node_shutdown(grpc_exec_ctx *exec_ctx, fd_node *fdn) {
+  gpr_mu_lock(&fdn->mu);
+  fdn->shutting_down = true;
+  if (!fdn->readable_registered && !fdn->writable_registered) {
+    gpr_mu_unlock(&fdn->mu);
+    fd_node_destroy(exec_ctx, fdn);
+  } else {
+    grpc_fd_shutdown(
+        exec_ctx, fdn->fd,
```

Comment:
can we create a local copy of `fdn->fd` so that we can release the lock before calling `grpc_fd_shutdown`?

Reply:
Thanks for the review!

We probably still need this lock when calling `grpc_fd_shutdown`. `on_readable/writable_cb` may be invoked by the polling engine at the same time we call `fd_node_shutdown`. If `fdn->mu` is released before `grpc_fd_shutdown`,  `fdn->fd` may get orphaned in  `on_readable/writable_cb` before `grpc_fd_shutdown`.

But `on_readable_cb` and `on_writable_cb` should make a local copy of `grpc_fd_wrapped_fd(fdn->fd)` in case `fdn` is freed in another thread after releasing the lock. I've fixed them.